src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/GlobalFlowStateAnalysis/GlobalFlowStateAnalysis.GlobalFlowStateAnalysisValueSetDomain.cs

```diff
@@ -79,7 +80,36 @@ public override GlobalFlowStateAnalysisValueSet Merge(GlobalFlowStateAnalysisVal
                 Debug.Assert(value1.Kind == GlobalFlowStateAnalysisValueSetKind.Known);
                 Debug.Assert(value2.Kind == GlobalFlowStateAnalysisValueSetKind.Known);
 
-                return new GlobalFlowStateAnalysisValueSet(value1, value2);
+                // Perform some early bail out checks.
+                if (Equals(value1, value2))
+                {
+                    return value1;
+                }
+
+                if (value1.Height == 0 && value1.AnalysisValues.IsSubsetOf(value2.AnalysisValues))
+                {
+                    return value2;
+                }
+
+                if (value2.Height == 0 && value2.AnalysisValues.IsSubsetOf(value1.AnalysisValues))
+                {
+                    return value1;
+                }
+
+                // Check if value1 and value2 are negations of each other.
+                // If so, the analysis values nullify each other and we return an empty set.
+                if (value1.Height == value2.Height &&
+                    value1.AnalysisValues.Count == value2.AnalysisValues.Count &&
+                    Equals(value1, value2.GetNegatedValue()))
```

Comment:
💡 Can you file a follow-up issue to investigate using a negating comparer here to avoid constructing a negated value every time?

Reply:
Sure will do